chainer/reporter.py

```diff
@@ -0,0 +1,325 @@
+        """Makes this reporter object current."""
+        _reporters.append(self)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Recovers the previous reporter object to the current."""
+        _reporters.pop()
+
+    @contextlib.contextmanager
+    def scope(self, observation):
+        """Creates a scope to report observed values to the given dictionary.
+
+        This is a context manager to be passed to ``with`` statements. In this
+        scope, the observation dictionary is changed to the given one.
+
+        It also makes this reporter object current.
+
+        Args:
+            observation (dict): Observation dictionary. All observations
+                reported inside of the ``with`` statement are written to this
+                dictionary.
+
+        """
+        old = self.observation
+        self.observation = observation
+        self.__enter__()
+        yield
+        self.__exit__(None, None, None)
+        self.observation = old
+
+    def add_observer(self, name, observer):
+        """Registers an observer of values.
+
+        Observer defines a scope of names for observed values. Values observed
+        with the observer are registered with names prefixed by the observer
+        name.
+
+        Args:
+            name (str): Name of the observer.
+            observer: The observer object. Note that the reporter distinguishes
+                the observers by their object ids (i.e., ``id(owner)``), rather
+                than the object equality.
+
+        """
+        self._observer_names[id(observer)] = name
+
+    def add_observers(self, prefix, observers):
+        """Registers multiple observers at once.
+
+        This is a convenient method to register multiple objects at once.
+
+        Args:
+            prefix (str): Prefix of each name of observers.
+            observers: Iterator of name and observer pairs.
+
+        """
+        for name, observer in observers:
+            self._observer_names[id(observer)] = prefix + name
+
+    def report(self, values, observer=None):
+        """Reports observed values.
+
+        The values are written with the key, prefixed by the name of the
+        observer object if given.
+
+        Args:
+            values (dict): Dictionary of observed values.
+            observer: Observer object. Its object ID is used to retrieve the
+                observer name, which is used as the prefix of the registration
+                name of the observed value.
+
+        """
+        if observer is not None:
+            observer_id = id(observer)
+            if observer_id not in self._observer_names:
+                raise KeyError(
+                    'Given observer is not registered to the reporter.')
+            observer_name = self._observer_names[observer_id]
+            for key, value in six.iteritems(values):
+                name = '%s/%s' % (observer_name, key)
+                self.observation[name] = value
```

Comment:
When the observer is not registered, it raises KeyError with the id of the observer. This error is difficult to interpret, so it is better to add a custom error message like "observer is not registered".